libbeat/esleg/eslegclient/connection.go

```diff
@@ -326,6 +326,9 @@ func (conn *Connection) Ping() (ESPingData, error) {
 // Close closes a connection.
 func (conn *Connection) Close() error {
 	conn.HTTP.CloseIdleConnections()
+	conn.cancelReqs()
+	// Creates a new context to be use in new requests
+	conn.reqsContext, conn.cancelReqs = context.WithCancel(context.Background())
```

Comment:
Doing this here feels weird. 

The actual problem feels like we are reusing a connection that is closed, which just seems wrong. 

Recreating the context is definitely the solution but this feels like the wrong place to be doing it, I'd expect it to happen when the connection is recreated.

Alternatively, we should just stop closing the connection if we are actually going to reuse it.

Reply:
I agree with everything you said @cmacknz, at least for this PR I didn't want to do a big refactoring as that's a pretty critical part of our code.

Looking at the code, this Connection type looks more like a low level client than an actual connection that has got a connected/disconnected state, hence the concept of closing it might not even make sense as it does not hold an actual active connection.

My suggestion: merge this PR without any architectural changes to make sure we're not "reintroducing" the issues fixed by https://github.com/elastic/beats/pull/40572 and create a follow up GH issue to rework this bit of the code base.
 

> I'd expect it to happen when the connection is recreated.

This connection is never recreated, that's the "problem", neither are the clients that use this connection. The [backoffClient](https://github.com/elastic/beats/blob/main/libbeat/outputs/backoff.go) is a good example (well, that's the one in my reproduction path) it "closes" the connection (in the snippet below client ends up using the Connection type we're talking about) on any publish error, but then Publish can just be called again (as it is) and it is expected to just work.
https://github.com/elastic/beats/blob/54c99614a62a1a704c8add8d5ee75e5189583926/libbeat/outputs/backoff.go#L60-L67

Reply:
Even with the explanation this doesn't look right to me. It's true the "connection" itself is only created once, but then `Close` / `Connect` are called repeatedly to cycle it on error. To me the minimal change to fix this particular issue is that the extra request context (which itself is quite new) should be created in `Connect` rather than in `NewConnection` as it currently is.

Reply:
Ok, I'll give it a try. I'll re-work the PR and see if I can get this context out of the connection creation

Reply:
@cmacknz, @faec, I've refactored it. There is a part of me a bit unease because I cannot say for sure that all users of `Connection` will always call `Connect` before using it... However the tests are passing and I didn't find any issues during manual testing.

Reply:
Done, it's ready for re-revewi